### DISCUSSION TITLE
OFF TOPIC - Some help/idea needed with a similar project

### DISCUSSION THREAD

**Discussion by GSzabados** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#457] [Show and tell]
@visualapproach, I am reaching out to you, because I am just starting to work on a quite similar project. I have an alarm panel which communicates in a similar way to the keypad attached to it. It has a DATA and CLK line to the keypad, which had 15 LEDs representing triggered sensors and zones and alarm states.

I have managed to hook up a Pololu 2595, 4-channel logic level shifter to these lines and record the signal with a logic level analyzer.

It looks like something like this.

![image](https://user-images.githubusercontent.com/35445496/221155440-f5aa27ac-b589-4ab3-9a5c-b6123c2d6a45.png)

My next step would be to decode the DATA to ones and zeros and then match each bit (bit pair) with the corresponding LED. I am not looking for control (yet), I just want to feed in the sensor state through MQTT to Home Assistant.

Could you please advise where to start looking at for the synchronous decoding (code) of the CLK and DATA with an ESP? The clock seems to be 2kHz, and the DATA can be all four possible state pairs during a clock period. The length of the signal is 32bits.

Any advise would be appreciated! Many thanks in advance!

**Comment by visualapproach** [comment]
Hi @GSzabados sure, I'm just confused about what you want to know. How to read the electrical signals with an ESP? Or how to decode the signals in theory?
/VA

**Comment by GSzabados** [comment]
I should have done a bit more research on your code, before asking this question. I had the impression, that you had to write from low-level up the whole communication of the device, but I just realised after looking at the code (finding the right file), that you do use a standard serial communication.

Unfortunately my alarm uses some old "proprietary" serialised communication. First I though, it might be I2C, but it isn't. Some things looks similar, but it is not I2C.

I have some ideas generally how to do this, I just miss the experience coding an ESP device to do it the right way. 😃

**Reply by visualapproach** [reply]
The easiest is if there is a standard format like 'serial'. As is the case for the 4-wire pumps. The 6-wire though has a SPI-like format. But not SPI. So then you have to 'bit bang' it yourself. I use interrupts on the clock line (and the CS but that's not relevant in your case) to know when to sample the data line, or write to it. If it's a bidirectional data line you have to switch pin mode to be able to write to it. Or it can be tackled by hardware in some cases.
Try to keep the interrupt routines as small as possible. Mine is stretching the limits I think.

**Reply by GSzabados** [reply]
Thanks! I will look at the 6-wire code as well. I looked at the 4-wire where I saw the serial communication.

It is SPI like as well, I would say. Because the LEDs are controlled by two HCF4021 (8-Stage Static Shift Register).

This is the keypad, which receives the DATA and CLK. It is really stretching my electronics knowledge. But the communication from the control panel should only set the LEDs, and that is what I am looking for really.

https://blog.uaid.net.ua/satel-ca-6-kled/

The text is incorrect, as I found the ICs are different on the picture than what is written in the text.

**Reply by visualapproach** [reply]
The text was cryptic to me... I'll be away for a day but you can see if you find some data sheets on those ICs. Sometime they reveal how to decode the signals. And without knowing, maybe the line is bidirectional- pressed buttons goes one way and LED-signals the other way.

**Reply by GSzabados** [reply]
I use Google Chrome, so right click and then there is an option "Translate to English"....

Thanks for the interrupt suggestion, that seems to be what I will need. I just want to read the DATA basically on (after) CHANGE of the CLK line.

The CLK is 2kHz, meanwhile 16 set of 32bits are transferred each second with big time gaps. (I think those are the windows when the keypad can communicate.) Hopefully it can fit into the interrupt.

**Reply by visualapproach** [reply]
2kHz is slow. The ESP can do 10000 operations for each bit so no need to optimize. Remember to put ISR in IRAM. And don't call other functions from the ISR if possible. I found that some versions of ESP32 is buggy with edge triggered interrupts. So maybe begin with 8266. That's just my experience 
Good luck now!

**Reply by GSzabados** [reply]
I can try with both. I have some at home from both of them. 

I really appreciate your help!

**Reply by GSzabados** [reply]
@visualapproach, thanks for the help!

I used at the end a M5-Stack ATOM Lite, which has an ESP32 PICO. It is all working. I managed to receive the state changes of the zones and send them to the MQTT broker.

https://github.com/GSzabados/SATEL_CA-6_to_MQTT

**Reply by visualapproach** [reply]
cool, god job!
